src/c++/perf_analyzer/metrics_manager.h

```diff
@@ -0,0 +1,100 @@
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <chrono>
+#include <condition_variable>
+#include <cstdint>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <utility>
+#include <vector>
+#include "client_backend/client_backend.h"
+#include "metrics.h"
+
+namespace triton { namespace perfanalyzer {
+
+#ifndef DOCTEST_CONFIG_DISABLE
+class TestMetricsManager;
+#endif
+
+class MetricsManager {
+ public:
+  MetricsManager(
+      std::shared_ptr<clientbackend::ClientBackend> client_backend,
+      uint64_t metrics_interval_ms);
+
+  /// Ends the background thread, redundant in case StopQueryingMetrics() isn't
+  /// called
+  ~MetricsManager();
+
+  /// Starts background thread that queries metrics on an interval
+  void StartQueryingMetrics();
+
+  /// Main loop of background thread that queries metrics on an interval
+  void QueryMetricsEveryNMilliseconds();
+
+  /// Checks if background thread threw exception and propogates it if so
+  void CheckQueryingStatus();
+
+  /// Swaps the the currently collected metrics with an empty external vector
+  void SwapMetrics(
+      std::vector<std::pair<
+          std::chrono::time_point<std::chrono::system_clock>, Metrics>>&
+          metrics_per_timestamp);
+
+  /// Ends the background thread
+  void StopQueryingMetrics();
+
+ private:
+  void CheckForMissingMetrics(const Metrics& metrics);
+  void CheckForMetricIntervalTooShort(
+      const std::chrono::nanoseconds& remainder,
+      const std::chrono::nanoseconds& duration);
+
+  std::shared_ptr<clientbackend::ClientBackend> client_backend_{nullptr};
+  uint64_t metrics_interval_ms_{0};
+  std::mutex metrics_per_timestamp_mutex_{};
+  std::vector<
+      std::pair<std::chrono::time_point<std::chrono::system_clock>, Metrics>>
```

Comment:
Can we make this structure simpler to include only a list of `Metrics`? I think the timestamp of when the Metrics were collected is not important to us.

Reply:
Good point. Now that metrics are associated with a PerfStatus, and that already tells us if it's in a stable pass, now we don't need the timestamp to distinguish which metrics were collected during stability passes. Removed the pair and now vector just holds the metrics 👍